pkg/sources/forward/data_forward.go

```diff
@@ -390,6 +433,31 @@ func (isdf *DataForward) forwardAChunk(ctx context.Context) {
 	metrics.ForwardAChunkProcessingTime.With(map[string]string{metrics.LabelVertex: isdf.vertexName, metrics.LabelPipeline: isdf.pipelineName, metrics.LabelVertexType: string(dfv1.VertexTypeSource), metrics.LabelVertexReplicaIndex: strconv.Itoa(int(isdf.vertexReplica))}).Observe(float64(time.Since(start).Microseconds()))
 }
 
+// isWatermarkIdle calculate that the current watermark is idling for the given maxWait duration based the provided
+// configuration by user for idle sourcing
+func (isdf *DataForward) isWatermarkIdle() bool {
+	if isdf.WatermarkConfig.IdleSource != nil {
+		// MaxWait is the wait time before publishing the idle watermark with MinIncrement value. If watermark found to be idle
+		// until MaxWait duration then publish the watermark by adding the MinIncrement value in it.
+		maxWait := isdf.WatermarkConfig.IdleSource.GetMaxWait()
+		if maxWait != nil && time.Since(isdf.lastUpdateWM) > *maxWait {
+			computedWM := isdf.wmFetcher.ComputeWatermark(nil, 0)
+			if isdf.currentWatermark == nil {
+				isdf.currentWatermark = &computedWM
+			}
+			if computedWM == *isdf.currentWatermark {
+				// set idleWMFound time if not already set and watermark is found to be idle
+				if isdf.idleWMFound.IsZero() {
+					isdf.idleWMFound = time.Now()
```

Comment:
isWatermarkIdle based on the name, is a READ(non-write) method which simply tells whether the source is idling. We have some write operations inside isWatermarkIdle which modify(WRITE) the states of the forwarder(L446, L451). Can we move them to the outside?

Reply:
I can move it to a different fn and call if after idle found